### PR TITLE
Remove await when(...) in Network and IBC vats

### LINKED ISSUE #8867: Convert awaits in network and ibc vats to promise watcher
## What is the Problem Being Solved?

We've been working on making the network and IBC vats durable (#8721). Leftover work from this initiative is that we haven't made promises durable yet, which prevents us from truly being able to upgrade.

## Description of the Design

As part of the `vow`(whenable) work, we've created promise watchers that are durable and can ensure that a promise is reconnected on upgrade. We need to go through the two vats and migrate any code awaiting on a Vow to use a watcher insead

open design questions:

 - [ ] use the vow-capable `E` work-alike?
 - [ ] convert _all_ awaits? or are there short-lived ones that should not be converted?

## Test Plan

 - [ ] upgrade / restart the vats at many points in the state machine

## Upgrade Considerations

```[tasklist]
### Tasks
- [ ] https://github.com/Agoric/agoric-sdk/issues/8906
- [ ] #9027 
```

### PR DESCRIPTION
**Pull request by iomekam**
closes https://github.com/Agoric/agoric-sdk/issues/8867

## Description

We've already done the work of removing all `FarRef` from the IBC and Network vats. The last remaining work to ensure these vats are durable is moving to a promise watcher pattern. We are already returning `Vow` where necessary, so this PR does the work of removing all `await when(...)`, and instead, use `watch`. This will allow promises to survive through upgrades. 

### Security Considerations

No new security assumptions/dependencies

### Scaling Considerations

Creation of a few more exo classes to power watching the promise in a durable way

### Documentation Considerations

Although these vats are not on-chain yet, we did make a change to following APIs to return a Vow instead of a promise:
- `Port.revoke`
- `BridgeManager.fromBridge`
- `ScopedBridgeManager.fromBridge`

### Testing Considerations

We've added a new testing suite to validate API behaviors before, during, and after an upgrade

### Upgrade Considerations

N/A

### CODE REVIEW

File: packages/vats/src/types.d.ts [hunk 1]
```diff
@@ -89,13 +89,13 @@ export type NamesByAddressAdmin = NameAdmin & {
 /** An object that can receive messages from the bridge device */
 export type BridgeHandler = {
   /** Handle an inbound message */
-  fromBridge: (obj: any) => Promise<void>;
+  fromBridge: (obj: any) => PromiseVow<void>;
```
Comment (dckc) [inline packages/vats/src/types.d.ts:90]: Is this a breaking change to the bridge vat interface? Are we upgrading the bridge vat and all its clients?

Comment (iomekam) [reply packages/vats/src/types.d.ts:90]: Great question. As it stands today, the `fromBridge` method relies on `ProtocolImpl`, which is created in the network vat. Given that the ibc calls the `inbound` method on `ProtocolImpl` and awaits its resolution, we do in fact need this to be a vow if we want this promise to survive an upgrade.

We currently didn't have any plans of upgrading the bridge vat and its client, so this is brings up an interesting intersection of how to proceed. The options I see are:
1. Go back to having `fromBridge` return a promise and use the `await when(...)`. This results in any `fromBridge` promise to be rejected on an IBC or Network vat upgrade. This _may_ be acceptable given that even with the change to have `fromBridge` return a `Vow`, the promise wouldn't be durable until we make corresponding changes in the `bridge` vat to properly watch the vow.
2. Do the work to upgrade the bridge vat and it's clients (as a follow up PR).

@michaelfig, thoughts here?

Comment (dckc) [reply packages/vats/src/types.d.ts:90]: Or maybe the interface can sort of vary by scoped bridge handler?
I expect expressing that as a static type could be hairy, but as an upgrade plan, maybe it works?

Comment (iomekam) [reply packages/vats/src/types.d.ts:90]: As an alternative to Option 2, we could do something like this: https://github.com/Agoric/agoric-sdk/pull/8964/commits/f002884446ecbf6cad1db64c6e61e17d7bc63f18, where we keep the `BridgeHandler` interface the same, but take advantage of `V`

Comment (michaelfig) [reply packages/vats/src/types.d.ts:90]: `PromiseVow<void>` is just `Promise<void | Vow<void>>`, which reduces to `Promise<void>`.  So you could have option 0:
0. Don't change anything in this PR, and just make a note that "if a given `fromBridge` returns an actual `Vow`, we will need to accommodate that in the `bridge` vat someday."

File: packages/vats/src/ibc.js [hunk 2]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+      attempt,
+      obj,
+      util,
+      asyncVersions,
+      rPortID,
+      rChannelID,
+      order,
+      hops,
+      version,
+    }),
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]: `zone.exo`'s have no local state.
```suggestion
```

File: packages/vats/src/ibc.js [hunk 3]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+    {
+      onFulfilled(attemptedLocal) {
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
      onFulfilled(attemptedLocal, watchContext) {
```

File: packages/vats/src/ibc.js [hunk 4]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+          attempt,
+          obj,
+          util,
+          asyncVersions,
+          rPortID,
+          rChannelID,
+          order,
+          hops,
+          version,
+        } = this.state;
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
        } = watchContext;
```

File: packages/vats/src/ibc.js [hunk 5]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+          // Clean up after our failed attempt.
+          channelKeyToAttempt.delete(channelKey);
+          channelKeyToInfo.delete(channelKey);
+          void E(attempt).close();
+          throw e;
+        }
+      },
+    },
+  );
+  return makeProtocolImplAttemptWatcher;
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
  return protocolImplAttemptWatcher;
```

File: packages/vats/src/ibc.js [hunk 6]
```diff
@@ -187,8 +362,17 @@ export const prepareIBCConnectionHandler = zone => {
...
+  const { makeVowKit, watch } = powers;
+
   const makeAckWatcher = prepareAckWatcher(zone);
+  const makeProtocolImplAttemptWatcher =
+    prepareProtocolImplAttemptWatcher(zone);
+  const makeProtocolImplInboundWatcher = prepareProtocolImplInboundWatcher(
+    zone,
+    powers,
+    makeProtocolImplAttemptWatcher,
+  );
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
  const protocolImplAttemptWatcher =
    makeProtocolImplAttemptWatcher(zone);
  const protocolImplInboundWatcher = makeProtocolImplInboundWatcher(
    zone,
    powers,
    protocolImplAttemptWatcher,
  );
```

File: packages/vats/src/ibc.js [hunk 7]
```diff
@@ -454,59 +636,21 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
...
                 ),
+                makeProtocolImplInboundWatcher({
+                  channelKeyToAttempt,
+                  channelKeyToInfo,
+                  obj,
+                  util,
+                  rPortID,
+                  rChannelID,
+                  hops,
+                }),
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
                protocolImplInboundWatcher,
                {
                  channelKeyToAttempt,
                  channelKeyToInfo,
                  obj,
                  util,
                  rPortID,
                  rChannelID,
                  hops,
                },
```

File: packages/vats/src/ibc.js [hunk 8]
```diff
@@ -624,8 +768,7 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
               const conn = channelKeyToConnP.get(channelKey);
               const data = base64ToBytes(data64);
 
-              watch(conn.send(data), makeAckWatcher(util, packet));
-              break;
+              return watch(conn.send(data), makeAckWatcher(util, packet));
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]: This `makeAckWatcher` could be a singleton as well.
```suggestion
              return watch(conn.send(data), ackWatcher, { util, packet });
```

File: packages/vats/src/ibc.js [hunk 9]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+        );
+      },
+    },
+  );
+  return makeProtocolImplInboundWatcher;
+};
+
+/** @param {import('@agoric/base-zone').Zone} zone */
+const prepareProtocolImplAttemptWatcher = zone => {
+  const makeProtocolImplAttemptWatcher = zone.exoClass(
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]: If we rely on `watch`'s propagating its third argument, `watchContext`, we only need a singleton watcher.
```suggestion
const makeProtocolImplAttemptWatcher = zone => {
  const protocolImplAttemptWatcher = zone.exo(
```

File: packages/vats/src/ibc.js [hunk 10]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<import('@agoric/vow').prepareVowTools>} powers
+ * @param {ReturnType<prepareProtocolImplAttemptWatcher>} makeProtocolImplAttemptWatcher
+ */
+const prepareProtocolImplInboundWatcher = (
+  zone,
+  { watch },
+  makeProtocolImplAttemptWatcher,
+) => {
+  const makeProtocolImplInboundWatcher = zone.exoClass(
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]: This watcher can be a singleton, too.

```suggestion
 * @param {ReturnType<makeProtocolImplAttemptWatcher>} protocolImplAttemptWatcher
 */
const makeProtocolImplInboundWatcher = (
  zone,
  { watch },
  protocolImplAttemptWatcher,
) => {
  const protocolImplInboundWatcher = zone.exo(
```

File: packages/vats/src/ibc.js [hunk 11]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+      hops,
+    }) => ({
+      channelKeyToAttempt,
+      channelKeyToInfo,
+      obj,
+      util,
+      rPortID,
+      rChannelID,
+      hops,
+    }),
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
```

File: packages/vats/src/ibc.js [hunk 12]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+    {
+      onFulfilled(attempt) {
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
      onFulfilled(attempt, watchContext) {
```

File: packages/vats/src/ibc.js [hunk 13]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+            attempt,
+            obj,
+            util,
+            asyncVersions,
+            rPortID,
+            rChannelID,
+            order,
+            hops,
+            version,
+          }),
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
          protocolImplAttemptWatcher,
          {
            ...obj,
            ...watchContext,
            attempt,
          },
```

File: packages/vats/src/ibc.js [hunk 14]
```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
...
+          channelKeyToInfo,
+          obj,
+          util,
+          rPortID,
+          rChannelID,
+          hops,
+        } = this.state;
+
+        const { channelID, portID, asyncVersions, order, version } = obj;
+
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
```

File: packages/network/src/network.js [hunk 15]
```diff
@@ -963,22 +1890,30 @@ export function prepareLoopbackProtocolHandler(zone, { when }) {
...
-        };
+          makeProtocolHandlerInstantiateWatcher(),
+        );
+
+        const instantiateVow = watch(
+          instantiateInnerVow,
+          makeRethrowUnlessMissingWatcher(),
+        );
+        return watch(
+          Promise.all([acceptVow, instantiateVow]),
```
Comment (michaelfig) [inline packages/network/src/network.js]: `Promise.all` does not work with Vows.  Maybe we should make one that does?

**Review by michaelfig — CHANGES_REQUESTED** [review]
Wow!  This is a lot of tough work that you managed to pull off.

My main concern is simplifying it and reducing lines of code to make `network.js` and `ibc.js` more understandable.

Sorry if my review comments come out of order.  https://github.com/Agoric/agoric-sdk/pull/8964#discussion_r1515249457 is a good starting point to understanding.

File: packages/boot/test/bootstrapTests/test-net-ibc-upgrade.ts [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+      return [label, opts];
+    },
+    sendPacket: async ([label, opts]) => {
+      await EV.sendOnly(opts.client).send(label);
+      return [label, opts];
+    },
+    respond: async ([label, opts]) => {
+      await EV.sendOnly(opts.server).dequeue('onReceive');
+      return [label, opts];
+    },
```
Comment (gibson042) [inline packages/boot/test/bootstrapTests/test-net-ibc-upgrade.ts]: Formally capturing something I had previously only said in more ephemeral channels: each `.dequeue(…)` should have its own step, to test for interruption of the affected promise.
```suggestion
    requestListening: async ([label, opts]) => {
      await EV.sendOnly(opts.server).listen();
      return [label, opts];
    },
    startListening: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onListen');
      return [label, opts];
    },
    startClient: async ([label, opts]) => {
      const started = await EV(zoe).startInstance(
        installation.ibcClientMock,
        {},
        {},
        { address: '/ibc-port/', networkVat },
      );
      t.truthy(started.creatorFacet, `${label} ibcClientMock`);
      return [label, { ...opts, client: started.creatorFacet }];
    },
    getAddresses: async ([label, opts]) => {
      const serverAddress = await EV(opts.server).getLocalAddress();
      const clientAddress = await EV(opts.client).getLocalAddress();
      t.log(`${label} server ${serverAddress} client ${clientAddress}`);
      return [label, { ...opts, serverAddress }];
    },
    requestConnection: async ([label, opts]) => {
      await EV.sendOnly(opts.client).connect(opts.serverAddress);
      return [label, opts];
    },
    acceptConnection: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onAccept');
      return [label, opts];
    },
    openConnection: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onOpen');
      return [label, opts];
    },
    sendPacket: async ([label, opts]) => {
      await EV.sendOnly(opts.client).send(label);
      return [label, opts];
    },
    respond: async ([label, opts]) => {
      await EV.sendOnly(opts.server).dequeue('onReceive');
      return [label, opts];
    },
```

File: packages/vats/src/ibc.js [hunk 17]
```diff
@@ -454,59 +426,22 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
...
                 /** @type {ProtocolImpl} */ (protocolImpl).inbound(
                   localAddr,
                   remoteAddr,
                 ),
+                this.facets.protocolImplInboundWatcher,
+                {
+                  channelKeyToAttempt,
+                  channelKeyToInfo,
+                  obj,
+                  util,
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]: These can be obtained from `this.state` and `this.facets.util`.
```suggestion
                  obj,
```

File: packages/vats/src/ibc.js [hunk 18]
```diff
@@ -624,8 +559,10 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
               const conn = channelKeyToConnP.get(channelKey);
               const data = base64ToBytes(data64);
 
-              watch(conn.send(data), makeAckWatcher(util, packet));
-              break;
+              return watch(conn.send(data), this.facets.ackWatcher, {
+                protocolUtils: util,
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]: Same.
```suggestion
```

File: packages/vats/src/ibc.js [hunk 19]
```diff
@@ -735,6 +672,103 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
           return kit;
         },
       },
+      ackWatcher: {
+        onFulfilled(ack, watcherContext) {
+          const { protocolUtils, packet } = watcherContext;
+
+          const realAck = ack || DEFAULT_ACKNOWLEDGEMENT;
+          const ack64 = dataToBase64(realAck);
+          protocolUtils
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
          const { packet } = watcherContext;

          const realAck = ack || DEFAULT_ACKNOWLEDGEMENT;
          const ack64 = dataToBase64(realAck);
          this.facets.util
```

File: packages/vats/src/ibc.js [hunk 20]
```diff
@@ -735,6 +672,103 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
...
+            attempt,
+            obj,
+            util,
+            asyncVersions,
+            rPortID,
+            rChannelID,
+            order,
+            hops,
+            version,
+          } = watcherContext;
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
            attempt,
            obj,
            asyncVersions,
            rPortID,
            rChannelID,
            order,
            hops,
            version,
          } = watcherContext;
          const {
            channelKeyToAttempt,
            channelKeyToInfo,
          } = this.state;
```

File: packages/vats/src/ibc.js [hunk 21]
```diff
@@ -735,6 +672,103 @@ export const prepareIBCProtocol = (zone, { makeVowKit, watch, when }) => {
...
+
+          channelKeyToAttempt.init(channelKey, attempt);
+          channelKeyToInfo.init(channelKey, obj);
+
+          const negotiatedVersion = match[3];
+
+          try {
+            if (asyncVersions) {
+              // We have async version negotiation, so we must call back now.
+              return util.downcall('tryOpenExecuted', {
```
Comment (michaelfig) [inline packages/vats/src/ibc.js]:
```suggestion
              return this.facets.util.downcall('tryOpenExecuted', {
```

File: packages/vow/src/watch-utils.js [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+ * @param {import('./watch.js').Watch} watch
+ * @param {() => import('./types.js').VowKit<any>} makeVowKit
+ */
+export const prepareWatchUtils = (zone, watch, makeVowKit) => {
+  const detached = zone.detached();
+  const makeWatchUtilsKit = zone.exoClassKit(
+    'WatchUtils',
+    {
+      utils: M.interface('Utils', {
+        all: M.call(M.arrayOf(VowShape)).returns(M.any()),
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]: `all` should accept and return arrays of any values (doesn't matter what they are), whether Vow, Promise, or any other passable.
```suggestion
        all: M.call(M.arrayOf(M.any())).returns(M.arrayOf(M.any())),
```

File: packages/vow/src/watch-utils.js [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+
+        /** @type {MapStore<bigint, number>} */
+        length: detached.mapStore('length'),
+
+        /** @type {MapStore<bigint, any[]>} */
+        results: detached.mapStore('results'),
+
+        /** @type {MapStore<bigint, import('./types.js').VowKit>} */
+        kits: detached.mapStore('kits'),
+      };
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]: I'd suggest grouping these stores together... I think the overhead is less that way:
```suggestion
    () => {
      /**
        * @typedef {object} VowState
        * @property {number} remaining
        * @property {MapStore<number, any>} resultMap
        * @property {import('./types.js').VowKit['resolver']} resolver
        */
        /** @type {MapStore<bigint, VowState>} */
        const idToVowState = detached.mapStore('idToVowState');
        return {
          nextId: 0n,
          idToVowState,
        };
      }  
```

File: packages/vow/src/watch-utils.js [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+          const results = this.state.results;
+
+          this.state.count.set(id, count);
+          results.set(id, [...results.get(id), value]);
+
+          if (count === length) {
+            kit.resolver.resolve([...results.get(id).values()]);
+          }
+        },
+      },
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]: We can inline the logic and avoid a separate `helpers` facet.
```suggestion
```

File: packages/vow/src/watch-utils.js [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+  const detached = zone.detached();
+  const makeWatchUtilsKit = zone.exoClassKit(
+    'WatchUtils',
+    {
+      utils: M.interface('Utils', {
+        all: M.call(M.arrayOf(VowShape)).returns(M.any()),
+      }),
+      helpers: M.interface('Helpers', {
+        check: M.call(M.any(), M.any()).returns(),
+      }),
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]: The watcher is the only other facet we need.
```suggestion
```

File: packages/vow/src/watch-utils.js [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+          this.state.count.init(id, 0);
+          this.state.length.init(id, vows.length);
+          this.state.results.init(id, harden([]));
+
+          for (const vow of vows) {
+            watch(vow, this.facets.watcher, { id });
+          }
+
+          this.state.id += 1n;
+          return kit.vow;
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]:
```suggestion
          const { nextId: id, idToVowState } = this.state;
          const kit = makeVowKit();

          // Preserve the order of the vow results.
          let index = 0;
          for (const vow of vows) {
            watch(vow, this.facets.watcher, { id, index });
            index += 1;
          }
          
          if (index > 0) {
            // Save the state until rejection or all fulfilled.
            this.state.nextId += 1n;
            idToVowState.init(id, harden({
              resolver: kit.resolver,
              remaining: index,
              resultsMap: detached.mapStore('resultsMap');
            });
         } else {
            // Base case: nothing to wait for.
            kit.resolver.resolve(harden([]));
         }
         return kit.vow;
```

File: packages/vow/src/watch-utils.js [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+            kit.resolver.resolve([...results.get(id).values()]);
+          }
+        },
+      },
+      watcher: {
+        onFulfilled(value, { id }) {
+          this.facets.helpers.check(value, id);
+        },
+        onRejected(value, { id }) {
+          this.facets.helpers.check(value, id);
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]: `onRejected` logic should just reject the vow:
```suggestion
          const { idToVowState } = this.state;
          if (!idToVowState.has(id)) {
            // First rejection wins.
            return;
          }
          const { resolver } = idToVowState.get(id);
          idToVowState.delete(id);
          resolver.reject(value);
```

File: packages/vow/src/watch-utils.js [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+
+          if (count === length) {
+            kit.resolver.resolve([...results.get(id).values()]);
+          }
+        },
+      },
+      watcher: {
+        onFulfilled(value, { id }) {
+          this.facets.helpers.check(value, id);
+        },
```
Comment (michaelfig) [inline packages/vow/src/watch-utils.js]:
```suggestion
        onFulfilled(value, { id, index }) {
          const { idToVowState } = this.state;
          if (!idToVowState.has(id)) {
            // Resolution of the returned vow happened already.
            return;
          }
          const { remaining, resultMap, resolver } = idToState.get(id);
          // Capture the fulfilled value.
          resultMap.init(index, value);
          const vowState = harden({
            remaining: remaining - 1,
            resultMap,
            resolver,
          });
          if (vowState.remaining > 0) {
            idToVowState.set(id, vowState);
            return;
          }
          // We're done!  Extract the array.
          idToVowState.delete(id);
          const results = new Array(resultMap.getSize());
          for (const [i, val] of resultMap.entries()) {
            results[i] = val;
          }
          resolver.resolve(harden(results));
        },
```

**Review by michaelfig — CHANGES_REQUESTED** [review]
Really close now!

The only rough edges that I saw were in `allVows`, so I made a lot of suggestions of what I'd like to see there.  Please expand the "Hidden conversations" to see all the details.

Ideally we'd have tests for `allVows` as well, but I'm okay with proceeding (I'll dismiss my "request changes" and do a quick "approve" stamp) once I see you've addressed my comments this round.

File: packages/vow/test/test-watch-utils.js [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+import { makeHeapZone } from '@agoric/base-zone/heap.js';
+
+import { prepareVowTools } from '../src/tools.js';
+
+test('allVows waits for a single vow to complete', async t => {
+  const zone = makeHeapZone();
+  const { watch, when, allVows } = prepareVowTools(zone);
+
+  const testPromiseP = Promise.resolve('promise');
+  const vowA = when(watch(testPromiseP));
```
Comment (michaelfig) [inline packages/vow/test/test-watch-utils.js]: For all of these input vows (like `vowA`, `vowB`, `vowC`), you should not need to use `when(watch(...))` as that returns a promise for the final result, never an intermediate vow.  Can you please change the input vows to be just `watch(testPromiseAP)` etc?

Do note that a `watch promises directly` test, such as `when(allVows(Promise.resolve('first'), Promise.resolve('second')))` is still valuable as a separate test, as well as `watch promises mixed with vows` is too.

**Review by michaelfig — APPROVED** [review]
Looking good!  I have just some suggestions for the `allVows` tests, then feel free to merge at your convenience.

**Pull request merged**